File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] toLeft = new int[nums.length];
        int[] toRight = new int[nums.length];
        int[] res = new int[nums.length];
        
        int product = 1;
        toLeft[0] = 1;
        toRight[nums.length - 1] = 1;
        
        for (int i = 1; i < nums.length; i++) {
            toLeft[i] = product * nums[i - 1];
            product *= nums[i - 1];
        }
        
        product = 1;
        
        for (int i = nums.length - 2; i >= 0; i--) {
            toRight[i] = product * nums[i+ 1];
            product *= nums[i + 1];
        }
        
        for (int i = 0; i < nums.length; i++) {
            res[i] = toLeft[i] * toRight[i];
        }
        
        return res;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = []
        n = len(nums)
        to_left = [1] * n
        to_right = [1] * n
        
        for i in range(n - 2, -1, -1):
            to_left[i] = to_left[i + 1] * nums[i + 1]
        
        for i in range(1, n, 1):
            to_right[i] = to_right[i - 1] * nums[i - 1]
            
        for i in range(n):
            res.append(to_left[i] * to_right[i])
            
        return res